keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+              mountPath: C:\csi
+            - name: csi-proxy-pipe
+              mountPath: \\.\pipe\csi-proxy
+      volumes:
+        - name: csi-proxy-pipe
+          hostPath: 
+            path: \\.\pipe\csi-proxy
+            type: null
+        - name: registration-dir
+          hostPath:
+            path: C:\var\lib\kubelet\plugins_registry\
+            type: Directory
+        - name: kubelet-dir
+          hostPath:
+            path: C:\var\lib\kubelet\
+            type: Directory
+        - name: test-dir
+          hostPath:
+            path: C:\temp\
+            type: DirectoryOrCreate
+        - name: plugin-dir
+          hostPath:
+            path: C:\var\lib\kubelet\plugins\pd.csi.storage.gke.io\
+            type: DirectoryOrCreate
+```
+
+Note that references to GCE PD CSI Plugin is used as an example above based on a prototype port of GCE PD CSI plugin with the enhancements in this KEP. Controller pods for the CSI plugin can be deployed on Linux nodes in the cluster in the same manner as it is done today.
+
+#### Deploy Windows workloads that consume persistent storage managed by a CSI plugin
+
+An operator should be able to deploy a Windows workload like SQL Server that consumes dynamically provisioned Persistent Volumes managed by a CSI plugin using:
+
+A storage class like:
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-gce-pd
+provisioner: pd.csi.storage.gke.io
+parameters:
+  type: pd-standard
+```
+
+with a PVC like:
+```
+apiVersion: v1
+metadata:
+  name: sqlpvc
+spec:
+  accessModes:
+    - ReadWriteOnce
+  storageClassName: csi-gce-pd
+  resources:
+    requests:
+      storage: 100Gi
+```
+
+and a Pod like:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: sqlserver1
+spec:
+  tolerations:
+  - key: "node.kubernetes.io/os"
+    operator: "Equal"
+    value: "win1809"
+    effect: "NoSchedule"
+  nodeSelector:
+    beta.kubernetes.io/os: windows
+  containers:
+  - name: sqlpod
+    image: ddebroy/sqlsrv:latest
+    volumeMounts:
+    - name: csi-sql-vol
+      mountPath: C:\Data
+    env:
+    - name: ACCEPT_EULA
+      value: "y"
+    - name: sa_password
+      value: "*****"
+  volumes:
+    - name: csi-sql-vol
+      persistentVolumeClaim:
+        claimName: sqlpvc
+```
+
+### Implementation Details
+
+CSI Node Plugins listen on domain sockets and respond to CSI API requests sent over GRPC from a container orchestrator like Kubernetes. They are responsible all storage management operations scoped around a specific node that are typically necessary after a CSI Controller Plugin has finished provisioning a Persistent Volume and attached it to the node. In Kubernetes, the CSI Node API calls are invoked by the CSI In-tree Plugin in the kubelet as well as the CSI Node Driver Registrar. The CSI Node Driver Registrar interacts with the Kubelet Plugin Watcher and it is maintained by the Kubernetes CSI community as a side-car container for deployment in CSI Node Plugin pods.
```

Comment:
For folks (like me) who are not familiar with the sockets used by CSI, it might be easier to digest if you list them here and state who creates and who use them and for what purposes.

Reply:
Does this diagram (already embedded in the KEP): https://ddebroywin1.s3-us-west-2.amazonaws.com/csi-proxy.png help?